### ISSUE TITLE
Shared Drives require "Manager" permission to work instead of "Content Manager" because of trashing restrictions

### ISSUE THREAD

**Issue by fxrobotinfo** [sabeechen/hassio-google-drive-backup#336]
Hello, I use Google Drive / shared drive / snapshot folder, and the addon after every file upload forget the snapshot folder (1VzKOvU4SYgM6f9f..........-vwl2U3I) code. If I put OK again for the same folder code, again working well - till not so long.
The code (1VzKOvU4SYgM6f9f..........-vwl2U3I) never change, but the system not accept without my next OK.
Maybe the others usually use the main GDrive, not with Shared Drive like me, and maybe there is the problem.
(two pictures attached) 
Thanks in advance. Kind Regards, Adam  

![gdrive_not-change-code-but-addon-forget-file-by-file](https://user-images.githubusercontent.com/42619538/105760776-89df4580-5f52-11eb-9be4-778318b3dde9.jpg)

![gdrive_backup_to_shared_gdrive-storage](https://user-images.githubusercontent.com/42619538/105761189-1ab62100-5f53-11eb-9238-5b6775ad9c99.jpg)

###### Description:
```
If you have anything else that could help explain what happened, click "Markdown" above and write it here.
```

 Addon version: `0.103.0`
 Home Assistant Version: `2021.1.4`
 Supervisor Version: `2021.01.7`
 Supervisor Channel: `stable`
 Hassos Version: `5.10`
 Docker Version: `19.03.13`
 Architecture: `aarch64`
 Machine: `raspberrypi4-64`
 Date: `2021-01-25T20:03:06.081595+00:00`
 Timezone: `Europe/Budapest`
 Failure Time: `2021-01-25T19:59:38.161396+00:00`
 Last Good Sync: `2021-01-25T20:00:06.861831+00:00`
 ###### Exception:
 ```
 No error could be identified automatically.
 ```
 Snapshots:
 ```
 {
    "HomeAssistant": {
        "snapshots": 9,
        "retained": 4,
        "deletable": 5,
        "name": "HomeAssistant",
        "title": "Home Assistant",
        "latest": "2021-01-25T19:18:13Z",
        "max": 5,
        "enabled": true,
        "size": "4.7 GB",
        "free_space": "34.4 GB"
    },
    "GoogleDrive": {
        "snapshots": 13,
        "retained": 2,
        "deletable": 11,
        "name": "GoogleDrive",
        "title": "Google Drive",
        "latest": "2021-01-25T19:18:13Z",
        "max": 50,
        "enabled": true,
        "size": "6.8 GB"
    }
}
 ```
 ###### Config:
 ```
 {
    "Setting.MAX_SNAPSHOTS_IN_HASSIO": 5,
    "Setting.MAX_SNAPSHOTS_IN_GOOGLE_DRIVE": 50,
    "Setting.DAYS_BETWEEN_SNAPSHOTS": 2.0,
    "Setting.SNAPSHOT_NAME": "REDACTED",
    "Setting.SNAPSHOT_TIME_OF_DAY": "23:00",
    "Setting.SNAPSHOT_PASSWORD": "REDACTED",
    "Setting.SPECIFY_SNAPSHOT_FOLDER": true,
    "Setting.GENERATIONAL_DAYS": 3,
    "Setting.GENERATIONAL_WEEKS": 4,
    "Setting.GENERATIONAL_MONTHS": 3,
    "Setting.GENERATIONAL_YEARS": 1,
    "Setting.GENERATIONAL_DAY_OF_MONTH": 28,
    "Setting.GENERATIONAL_DAY_OF_YEAR": 364
}
 ```
 ###### Addon Logs:
 ```
 01-25 20:57:16 INFO [backup.drive.folderfinder] Provided snapshot folder 1VzKOvU4SYgM6f9f..........-vwl2U3I is invalid
01-25 20:57:16 ERROR [backup.model.coordinator] The choosen backup folder has become inaccessible.  Please visit the addon web UI to select a backup folder.
01-25 20:57:16 INFO [backup.model.coordinator] I'll try again in 20 seconds
01-25 20:57:36 DEBUG [backup.model.syncer] Sync requested by Coordinator
01-25 20:57:36 INFO [backup.model.coordinator] Syncing Snapshots
01-25 20:57:36 DEBUG [backup.ha.harequests] Making Hassio request: http://hassio/snapshots
01-25 20:57:36 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/drive/v3/files/1VzKOvU4SYgM6f9f..........-vwl2U3I/?fields=id%2Cname%2CappProperties%2Csize%2Ctrashed%2CmimeType%2CmodifiedTime%2Ccapabilities%2Cparents&supportsAllDrives=true
01-25 20:57:36 INFO [backup.drive.folderfinder] Provided snapshot folder 1VzKOvU4SYgM6f9f..........-vwl2U3I is invalid
01-25 20:57:36 ERROR [backup.model.coordinator] The choosen backup folder has become inaccessible.  Please visit the addon web UI to select a backup folder.
01-25 20:57:36 INFO [backup.model.coordinator] I'll try again in 40 seconds
01-25 20:58:17 DEBUG [backup.model.syncer] Sync requested by Coordinator
01-25 20:58:17 INFO [backup.model.coordinator] Syncing Snapshots
01-25 20:58:17 DEBUG [backup.ha.harequests] Making Hassio request: http://hassio/snapshots
01-25 20:58:17 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/drive/v3/files/1VzKOvU4SYgM6f9f..........-vwl2U3I/?fields=id%2Cname%2CappProperties%2Csize%2Ctrashed%2CmimeType%2CmodifiedTime%2Ccapabilities%2Cparents&supportsAllDrives=true
01-25 20:58:17 INFO [backup.drive.folderfinder] Provided snapshot folder 1VzKOvU4SYgM6f9f..........-vwl2U3I is invalid
01-25 20:58:17 ERROR [backup.model.coordinator] The choosen backup folder has become inaccessible.  Please visit the addon web UI to select a backup folder.
01-25 20:58:17 INFO [backup.model.coordinator] I'll try again in 1(ish) minutes
01-25 20:59:37 DEBUG [backup.model.syncer] Sync requested by Coordinator
01-25 20:59:37 INFO [backup.model.coordinator] Syncing Snapshots
01-25 20:59:37 DEBUG [backup.ha.harequests] Making Hassio request: http://hassio/snapshots
01-25 20:59:37 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/drive/v3/files/1VzKOvU4SYgM6f9f..........-vwl2U3I/?fields=id%2Cname%2CappProperties%2Csize%2Ctrashed%2CmimeType%2CmodifiedTime%2Ccapabilities%2Cparents&supportsAllDrives=true
01-25 20:59:38 INFO [backup.drive.folderfinder] Provided snapshot folder 1VzKOvU4SYgM6f9f..........-vwl2U3I is invalid
01-25 20:59:38 ERROR [backup.model.coordinator] The choosen backup folder has become inaccessible.  Please visit the addon web UI to select a backup folder.
01-25 20:59:38 INFO [backup.model.coordinator] I'll try again in 2(ish) minutes
01-25 21:00:06 DEBUG [backup.ha.harequests] Making Hassio request: http://hassio/addons/self/options
01-25 21:00:06 INFO [backup.drive.folderfinder] Saving snapshot folder: 1VzKOvU4SYgM6f9f..........-vwl2U3I
01-25 21:00:06 DEBUG [backup.model.syncer] Sync requested by UI Server
01-25 21:00:06 INFO [backup.model.coordinator] Syncing Snapshots
01-25 21:00:06 DEBUG [backup.ha.harequests] Making Hassio request: http://hassio/snapshots
01-25 21:00:06 DEBUG [backup.drive.driverequests] Making Google Drive request: https://www.googleapis.com/drive/v3/files/?q=%271VzKOvU4SYgM6f9f..........-vwl2U3I%27+in+parents&fields=nextPageToken%2Cfiles%28id%2Cname%2CappProperties%2Csize%2Ctrashed%2CmimeType%2CmodifiedTime%2Ccapabilities%2Cparents%29&pageSize=100&supportsAllDrives=true&includeItemsFromAllDrives=true&corpora=allDrives
 ```
 ###### Supervisor Logs:
 ```
 [32m21-01-25 19:29:23 INFO (SyncWorker_7) [supervisor.snapshots.snapshot] Snapshot folder ssl[0m
[32m21-01-25 19:29:23 INFO (SyncWorker_7) [supervisor.snapshots.snapshot] Snapshot folder ssl done[0m
[32m21-01-25 19:29:23 INFO (SyncWorker_1) [supervisor.snapshots.snapshot] Snapshot folder addons/local[0m
[32m21-01-25 19:29:23 INFO (SyncWorker_1) [supervisor.snapshots.snapshot] Snapshot folder addons/local done[0m
[32m21-01-25 19:29:23 INFO (SyncWorker_3) [supervisor.snapshots.snapshot] Snapshot folder share[0m
[32m21-01-25 19:29:23 INFO (SyncWorker_3) [supervisor.snapshots.snapshot] Snapshot folder share done[0m
[32m21-01-25 19:29:23 INFO (SyncWorker_0) [supervisor.snapshots.snapshot] Snapshot folder homeassistant[0m
[32m21-01-25 19:33:01 INFO (MainThread) [supervisor.api.security] /snapshots access from cebe7a76_hassio_google_drive_backup[0m
[32m21-01-25 19:34:37 INFO (SyncWorker_0) [supervisor.snapshots.snapshot] Snapshot folder homeassistant done[0m
[32m21-01-25 19:35:29 INFO (MainThread) [supervisor.snapshots] Creating full-snapshot with slug 13bb120a completed[0m
[32m21-01-25 19:35:29 INFO (MainThread) [supervisor.api.security] /snapshots access from cebe7a76_hassio_google_drive_backup[0m
[32m21-01-25 19:35:29 INFO (MainThread) [supervisor.api.security] /snapshots/13bb120a/info access from cebe7a76_hassio_google_drive_backup[0m
[32m21-01-25 19:35:33 INFO (MainThread) [supervisor.api.security] /snapshots access from cebe7a76_hassio_google_drive_backup[0m
[32m21-01-25 19:38:42 INFO (MainThread) [supervisor.homeassistant.api] Updated Home Assistant API token[0m
[32m21-01-25 19:53:38 INFO (MainThread) [supervisor.api.security] /snapshots access from cebe7a76_hassio_google_drive_backup[0m
[32m21-01-25 19:53:39 INFO (MainThread) [supervisor.api.security] /snapshots/6df0ef34/remove access from cebe7a76_hassio_google_drive_backup[0m
[32m21-01-25 19:53:39 INFO (MainThread) [supervisor.snapshots] Removed snapshot file 6df0ef34[0m
[32m21-01-25 19:53:39 INFO (MainThread) [supervisor.api.security] /snapshots/13bb120a/download access from cebe7a76_hassio_google_drive_backup[0m
[32m21-01-25 19:53:39 INFO (MainThread) [supervisor.api.snapshots] Downloading snapshot 13bb120a[0m
[32m21-01-25 19:57:04 INFO (MainThread) [supervisor.api.security] /snapshots access from cebe7a76_hassio_google_drive_backup[0m
[32m21-01-25 19:57:05 INFO (MainThread) [supervisor.api.security] /snapshots access from cebe7a76_hassio_google_drive_backup[0m
[32m21-01-25 19:57:16 INFO (MainThread) [supervisor.api.security] /snapshots access from cebe7a76_hassio_google_drive_backup[0m
[32m21-01-25 19:57:36 INFO (MainThread) [supervisor.api.security] /snapshots access from cebe7a76_hassio_google_drive_backup[0m
[32m21-01-25 19:58:17 INFO (MainThread) [supervisor.api.security] /snapshots access from cebe7a76_hassio_google_drive_backup[0m
[32m21-01-25 19:59:37 INFO (MainThread) [supervisor.api.security] /snapshots access from cebe7a76_hassio_google_drive_backup[0m
[32m21-01-25 20:00:06 INFO (MainThread) [supervisor.api.security] /snapshots access from cebe7a76_hassio_google_drive_backup[0m
[32m21-01-25 20:02:37 INFO (MainThread) [supervisor.resolution.check] Starting system checks with state CoreState.RUNNING[0m
[32m21-01-25 20:02:37 INFO (MainThread) [supervisor.resolution.check] System checks complete[0m
[32m21-01-25 20:03:06 INFO (MainThread) [supervisor.api.security] /supervisor/logs access from cebe7a76_hassio_google_drive_backup[0m

 ```
 ###### Home Assistant Core Logs:
 ```
 [31m2021-01-25 20:57:04 ERROR (Bot:1044052683:updater) [telegram.ext.updater] Error while getting Updates: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running[0m
[31m2021-01-25 20:57:04 ERROR (Bot:1044052683:dispatcher) [homeassistant.components.telegram_bot.polling] Update "None" caused error: "Conflict: terminated by other getUpdates request; make sure that only one bot instance is running"[0m
[33m2021-01-25 20:57:10 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: synology_dsm[0m
[36m2021-01-25 20:57:32 DEBUG (MainThread) [custom_components.hacs] Nothing in the queue[0m
[31m2021-01-25 20:57:38 ERROR (Bot:1044052683:updater) [telegram.ext.updater] Error while getting Updates: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running[0m
[31m2021-01-25 20:57:38 ERROR (Bot:1044052683:dispatcher) [homeassistant.components.telegram_bot.polling] Update "None" caused error: "Conflict: terminated by other getUpdates request; make sure that only one bot instance is running"[0m
[33m2021-01-25 20:58:10 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: synology_dsm[0m
[31m2021-01-25 20:58:12 ERROR (Bot:1044052683:updater) [telegram.ext.updater] Error while getting Updates: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running[0m
[31m2021-01-25 20:58:12 ERROR (Bot:1044052683:dispatcher) [homeassistant.components.telegram_bot.polling] Update "None" caused error: "Conflict: terminated by other getUpdates request; make sure that only one bot instance is running"[0m
[31m2021-01-25 20:58:45 ERROR (Bot:1044052683:updater) [telegram.ext.updater] Error while getting Updates: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running[0m
[31m2021-01-25 20:58:45 ERROR (Bot:1044052683:dispatcher) [homeassistant.components.telegram_bot.polling] Update "None" caused error: "Conflict: terminated by other getUpdates request; make sure that only one bot instance is running"[0m
[33m2021-01-25 20:59:10 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: synology_dsm[0m
[31m2021-01-25 20:59:18 ERROR (Bot:1044052683:updater) [telegram.ext.updater] Error while getting Updates: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running[0m
[31m2021-01-25 20:59:18 ERROR (Bot:1044052683:dispatcher) [homeassistant.components.telegram_bot.polling] Update "None" caused error: "Conflict: terminated by other getUpdates request; make sure that only one bot instance is running"[0m
[31m2021-01-25 20:59:52 ERROR (Bot:1044052683:updater) [telegram.ext.updater] Error while getting Updates: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running[0m
[31m2021-01-25 20:59:52 ERROR (Bot:1044052683:dispatcher) [homeassistant.components.telegram_bot.polling] Update "None" caused error: "Conflict: terminated by other getUpdates request; make sure that only one bot instance is running"[0m
[33m2021-01-25 21:00:10 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: synology_dsm[0m
[31m2021-01-25 21:00:25 ERROR (Bot:1044052683:updater) [telegram.ext.updater] Error while getting Updates: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running[0m
[31m2021-01-25 21:00:25 ERROR (Bot:1044052683:dispatcher) [homeassistant.components.telegram_bot.polling] Update "None" caused error: "Conflict: terminated by other getUpdates request; make sure that only one bot instance is running"[0m
[31m2021-01-25 21:00:59 ERROR (Bot:1044052683:updater) [telegram.ext.updater] Error while getting Updates: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running[0m
[31m2021-01-25 21:00:59 ERROR (Bot:1044052683:dispatcher) [homeassistant.components.telegram_bot.polling] Update "None" caused error: "Conflict: terminated by other getUpdates request; make sure that only one bot instance is running"[0m
[33m2021-01-25 21:01:10 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: synology_dsm[0m
[31m2021-01-25 21:01:32 ERROR (Bot:1044052683:updater) [telegram.ext.updater] Error while getting Updates: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running[0m
[31m2021-01-25 21:01:32 ERROR (Bot:1044052683:dispatcher) [homeassistant.components.telegram_bot.polling] Update "None" caused error: "Conflict: terminated by other getUpdates request; make sure that only one bot instance is running"[0m
[31m2021-01-25 21:02:06 ERROR (Bot:1044052683:updater) [telegram.ext.updater] Error while getting Updates: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running[0m
[31m2021-01-25 21:02:06 ERROR (Bot:1044052683:dispatcher) [homeassistant.components.telegram_bot.polling] Update "None" caused error: "Conflict: terminated by other getUpdates request; make sure that only one bot instance is running"[0m
[33m2021-01-25 21:02:10 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: synology_dsm[0m
[31m2021-01-25 21:02:39 ERROR (Bot:1044052683:updater) [telegram.ext.updater] Error while getting Updates: Conflict: terminated by other getUpdates request; make sure that only one bot instance is running[0m
[31m2021-01-25 21:02:39 ERROR (Bot:1044052683:dispatcher) [homeassistant.components.telegram_bot.polling] Update "None" caused error: "Conflict: terminated by other getUpdates request; make sure that only one bot instance is running"[0m

 ```

**Comment by fxrobotinfo** [comment]
Addon version: 0.103.0  »»» I refreshed 0.103.1, but the problem further I have.

**Comment by sabeechen** [comment]
I think you're probably right, the issue is most likely with shared drives, which are uncommon for people to use.  The permission model for them in Drive is very different and changes often.  I'll need to create a Google Apps domain and do some testing.

**Comment by sabeechen** [comment]
A couple questions to help me debug this issue:
* How long does it usually take for the error to come up after resaving the folder?
* What permissions does your user have on the Shared Drive and the snapshot folder?  This would be the user you originally used to authorize the addon with google drive.
* Do you have any restrictions set up on your Google Workspace account to limit sharing settings (eg restricted sharing outside your domain, etc)?

**Comment by sabeechen** [comment]
I think I figured out whats going on here.

TL;DR

- I suspect the Google Drive user you have the addon linked to is a "Content Manager" on the shared drive instead of a "Manager".
- "Content Manager" isn't able to permanently delete files in a shared drive, which makes the addon upset.
- Make your user a "Manager" on the shared drive to fix the problem for now.
- In the near future, I plan to make the addon work with the "Content Manager" role (with some caveats)

Long winded answer:
After doing a little more digging, I suspect this issue you're having is with the "Content Manager" role.  When I create a shared drive where my user has the "Content Manager" role, I run into the same problem you're seeing.  This role is relatively new, so I had to do some research to figure out why its causing problems.

Based on Google's docs about [shared drive access permissions](https://support.google.com/a/answer/7337554?hl=en), the Content Manager role doesn't have permission to _"Permanently delete files and folders in the trash"_.  Its necessary for the addon to have permission to completely delete files because otherwise it can only trash snapshots, which still takes up quota.

However, maybe that doesn't make sense for shared drives.  According to Google support, there is no space usage limit for a shared drive.  Interestingly, I wasn't able to find any reference to this specifically in [Google's documentation](https://support.google.com/a/answer/7338880?hl=en) and had to contact customer support to be told _"They do not have any storage limit. However, there are limitations to the number of files and uploads per day."_  Are they trying to hide this, or is it likely to change in the future?

I think what I'm going to do is that if snapshots are stored in a shared drive and user doesn't have permission to permanently delete but does have permission to trash items (eg the content manager role), then the addon will put old snapshots in the trash instead of deleting them permanently.  Since shared drives have no storage limit, this shouldn't put anyone over a quota limit and they'll get deleted automatically after the trash is purged in 30 days.  I'll probably add a dialog that explains this if someone chooses to use the addon in this way.
 
In the meantime if you make the user you linked the addon to a "Manager" instead of a "Content Manager" on the shared drive, it should mitigate this problem  until I can make a new release with the changes.

**Comment by fxrobotinfo** [comment]
My shared (unlimited) drive I purchased from Aliexpress, and from one months ago till now was empty two times my purchased drive - I was not happy, the "seller" between many sorry say never more will be empty all of my saved data. 
I dont have grant permission for this unlimited shared drive, but I can use and that true no volume limit. 
Now I just testing this snapshot save method, and I make other place also snapshot. 
Just for note: this unlimited shared drive I use about from one month ago, but the folder code forgetting I hope I remember good, but cca two weeks ago have, before was okay the folder code.
Have a good work.

**Comment by sabeechen** [comment]
I'm having a little trouble understanding what you're saying, but if I interpret correctly:

- You bought the shared drive off Aliexpress
- You had everything deleted from it a few times, which you've sorted out with the seller.
- You are saving your snapshots somewhere else, but just wanted to try putting them in a shared drive.
- The shared drive folder worked with the addon in the past.

If I misunderstand you can write your response in your native language and I can use Google Translate on it which might make it easier for me to get what you're saying (sorry, I only know English, Spanish, and a little Japanese)

Some answers:
- I can't be certain why the folder worked in the past, but it seems likely that the seller changed your permission on the Shared Drive at some time.  If you were a "Manager" before and only a "Content Manager" now, that would explain why it stopped working.
- I'll be making a release soon with changes that will make the "Content Manager" role work, but I can't promise exactly when that will be.
- You might want to be careful with this Shared Drive.  What the seller is doing violates Google's terms of service, so its likely they'll get shut down in the future if Google takes notice.  Additionally the seller is the "Administrator" of the shared drive and has access to all the file you put in it.  Their postings on aliexpress are not entirely honest about what they're doing.

**Comment by fxrobotinfo** [comment]
Yes, you understand good my wrong english. 
Thanks for your answer, and I let you know I use the snapshot with secret pw save.
I wait your update, with normal GDrive account the app for me till now also absolute perfect.

**Comment by sabeechen** [comment]
It took me a longer than I hoped to release this feature, but v0.104.1 is available now with this functionality.
To update the addon, go to:
`Supervisor` > `Add-on Store` > Click the `...` in the top right > `Refresh`
and you should see the new version.

Kindly reopen this bug and let me know if you continue to run into problems after updating.

**Comment by fxrobotinfo** [comment]
Hello Maistro, the issue is right to close now, because my problem absolutely resolved, I use the same shared GDrive and work well all of my snapshot save. Thanks again

**Comment by sabeechen** [comment]
Its worth mentioning that when the addon is saving snapshots into a Shared Drive as a Content Manager (eg what you're doing) then when old snapshots need to get removed to make space they'll be sent to the trash rather than deleted.  This is because the Content Manager doesn't have permission to permanently delete files from a Shared Drive.  Once they're in the trash they'll get automatically deleted permanently in 30 days.  I don't think this will be a problem because Shared Drives have unlimited storage.
